README.md

```diff
@@ -17,3 +17,379 @@ Implement a prototype job worker service that provides an API to run arbitrary L
+
+### API Design
+
+* The client will have an Command Api that takes in three types of request, Start, Stop, and Query
+
+* The client can then execute the commands over the server via Execute APIs'
+
+#### Start
+
+* The Start command is called with a StartRequest that has the client's command and required arguments and optional env, dir params
+
+* A uuid (universal unique identification) will be generated and a folder called START-\<startTimeStamp\>will be created, two logs called stdout.log and stderr.log will be created
+
+* the start command will execute the job and return with the uuid, pid, startTimeStamp, if it fails the process table will be correspondingly updated
+
+* goroutines should manage running processes in the background (outputing into logs, updating dataStore)
+
+* when the job is done the process table will be updated
+
+```
+type StartRequest {
+    //path of a program or executable
+    string command
+    //arguments to invoke program or executable with
+    []string args 
+    // environment variables for the execution
+    //OPTIONAL variable
+    []string env
+    // current working directory of the execution
+    //OPTIONAL variable, default will be the working directory of server 
+    string dir 
+}
+
+type StartResponse {
+    //process Id of the running process executed by the command
+    //will be 0 if process failed to execute
+    int pid 
+    //univeral unique identifier that tags each unique request made to server
+    string uuid
+    //starting time of start request
+    string startTimeStamp
+    string status
+    int statusCode
+    
+
+}
+
+func Start(StartRequest) returns(StartResponse)
+```
+
+#### Stop
+
+* The User should be able to stop the request based on the pid
+
+* When stopped the pid should be killed 
+
+* A response will be sent to client indicating process have been stopped along with the contents of the log
+
+* job should be marked as completed with the exit code in the dataStore
+
+```
+type StopRequest {
+    string uuid
+}
+
+type StopResponse {
+    []byte stdout
+    []byte stderr
+    bool isKilled
+    string endTimeStamp
+    string status
+    int statusCode
+}
+
+func Stop(StopRequest) returns(StopResponse)
+
+```
+
+#### Query 
+
+
+* There will be only two types of Query
+
+    * QueryOneProcess:
+
+        * return the logs of a job using a valid pid (if it was started) or uuid, along with ProcessInfo
+
+
+    * QueryRunningProcesses:
+
+        * get a list of job's executed and processInfo for the jobs
+
+```
+type ProcessInfo {
+    int pid 
+    string startTimeStamp 
+    string endTimeStamp
+    string processName 
+    bool isRunning 
+    int exitCode
+    string status
+}
+
+type QueryOneProcessRequest {
+    string uuid
+}
+
+type QueryOneProcessResponse {
+    processInfo procInfo
+    []byte stdout 
+    []byte stderr 
+    string status
+    int statusCode
```

Comment:
if these represent the status of request processing (success/failure), then they should be returned as gRPC errors instead